### PR TITLE
refactor: manage Artifactory credentials via context manager

### PR DESCRIPTION
**Pull request by rnetser**
##### What this PR does / why we need it:
Replace Artifactory Secret/ConfigMap `get_*` + `cleanup_*` pairs with an `artifactory_credentials` context manager (`ArtifactoryCredentials` dataclass) so lifecycle is tied to `with` / `ExitStack` and resources cannot leak on partial create failure.

`client: DynamicClient` is **mandatory** on `artifactory_credentials`, `get_artifactory_secret`, and `get_artifactory_config_map` (no implicit `get_client()`). Call sites pass an explicit client. `create_dv` composes selective create via ExitStack (`create_secret` / `create_config_map`). Unit tests cover the new API and `create_dv` wiring.

Assisted-by: Cursor Grok 4.5 <noreply@cursor.com>
Signed-off-by: rnetser <rnetser@redhat.com>

##### Special notes for reviewer:
- Fixed-name Secret/ConfigMap delete-on-exit (even if pre-existing) matches prior get+cleanup semantics intentionally.
- `.flake8` adds `enter_context` to `fcn_exclude_functions` so `ExitStack.enter_context(...)` is allowed (project FCN exception list; not a `# noqa`).
- `utilities/storage.py` is coverage-omitted; added focused `TestCreateDvArtifactory` unit tests for the new ExitStack/artifactory paths.

##### jira-ticket:
NONE

## Summary by CodeRabbit

* **Refactor**
  * Standardized Artifactory credential handling across VM, DataVolume, migration, backup, and chaos workflows.
  * Improved automatic cleanup while preserving existing credentials and resources.
  * Simplified registry-backed test configuration with consistent credential reuse.
  * Improved reliability when creating resources with missing or partially configured credentials.

* **Tests**
  * Expanded coverage for credential creation, reuse, selective provisioning, cleanup, and failure handling.
  * Updated workflow validation for consolidated credentials and dependable resource cleanup.

### CODE REVIEW

**Comment by openshift-virtualization-qe-bot-3** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

## Welcome! 🎉

This pull request will be automatically processed with the following features:

### 🔄 Automatic Actions
* **Reviewer Assignment**: Reviewers are automatically assigned based on the OWNERS file in the repository root
* **Size Labeling**: PR size labels (XS, S, M, L, XL, XXL) are automatically applied based on changes
* **Issue Creation**: A tracking issue is created for this PR and will be closed when the PR is merged or closed
* **Branch Labeling**: Branch-specific labels are applied to track the target branch
* **Auto-verification**: Auto-verified users have their PRs automatically marked as verified
* **Labels**: Enabled categories: `branch`, `can-be-merged`, `cherry-pick`, `has-conflicts`, `hold`, `needs-rebase`, `size`, `verified`, `wip`

### 📋 Available Commands

#### PR Status Management
* `/wip` - Mark PR as work in progress (adds WIP: prefix to title)
* `/wip cancel` - Remove work in progress status
* `/hold` - Block PR merging (PR author or approvers)
* `/hold cancel` - Unblock PR merging (PR author or approvers)
* `/verified` - Mark PR as verified
* `/verified cancel` - Remove verification status
* `/reprocess` - Trigger complete PR workflow reprocessing (useful if webhook failed or configuration changed)
* `/regenerate-welcome` - Regenerate this welcome message
* `/security-override` - Set security check runs to pass (maintainers only)
* `/security-override cancel` - Re-run security checks

#### Review & Approval
* `/lgtm` - Approve changes (looks good to me)
* `/approve` - Approve PR (approvers only)
* `/assign-reviewers` - Assign reviewers based on OWNERS file
* `/assign-reviewer @username` - Assign specific reviewer
* `/check-can-merge` - Check if PR meets merge requirements

#### Testing & Validation
 * `/retest tox` - Run Python test suite with tox
 * `/retest build-container` - Rebuild and test container image
 * `/retest verify-bugs-are-open` - verify-bugs-are-open
 * `/retest all` - Run all available tests


#### Container Operations
* `/build-and-push-container` - Build and push container image (tagged with PR number)
  * Supports additional build arguments: `/build-and-push-container --build-arg KEY=value`

#### Cherry-pick Operations
* `/cherry-pick <branch>` - Schedule cherry-pick to target branch when PR is merged
  * Multiple branches: `/cherry-pick branch1 branch2 branch3`
* `/cherry-pick-retry <branch>` - Retry a failed cherry-pick (merged PRs only)

#### Branch Management
* `/rebase` - Rebase this PR branch onto its base branch

#### Label Management
* `/<label-name>` - Add a label to the PR
* `/<label-name> cancel` - Remove a label from the PR

### ✅ Merge Requirements

This PR will be automatically approved when the following conditions are met:

1. **Approval**: `/approve` from at least one approver
2. **LGTM Count**: Minimum 2 `/lgtm` from reviewers
3. **Status Checks**: All required status checks must pass
4. **No Blockers**: No wip, hold, has-conflicts labels and PR must be mergeable (no conflicts)
5. **Verified**: PR must be marked as verified

### 📊 Review Process

<details>
<summary><strong>Approvers and Reviewers</strong></summary>


**Approvers:**
 * dshchedr
 * jpeimer
 * myakove
 * rnetser
 * vsibirsk


**Reviewers:**
 * Acedus
 * Ahmad-Hafe
 * Dsanatar
 * OhadRevah
 * RoniKishner
 * SamAlber
 * acinko-rh
 * akalenyu
 * akri3i
 * albarker-rh
 * awels
 * dalia-frank
 * dshchedr
 * ema-aka-young
 * geetikakay
 * guchen11
 * hmeir
 * josemacassan
 * jpeimer
 * kgoldbla
 * kshvaika
 * mijankow
 * rlobillo
 * rnetser
 * sarahbx
 * vsibirsk


</details>

<details>
<summary><strong>Available Labels</strong></summary>

 * hold
 * verified
 * wip
 * lgtm
 * approve

</details>

<details>
<summary><strong>AI Features</strong></summary>

* **Cherry-Pick Conflict Resolution**: Enabled (claude/claude-opus-4-6-1m)

</details>

<details>
<summary><strong>Security Checks</strong></summary>

* **Suspicious Path Detection**: Monitors paths: `.claude/`, `.vscode/`, `.cursor/`, `.devcontainer/`, `.pi/`, `.github/workflows/`, `.github/actions/`
* **Committer Identity Check**: Verifies last committer matches PR author
* **Mandatory**: Security checks block merge (use `/security-override` to bypass — maintainers only)

</details>

### 💡 Tips

* **WIP Status**: Use `/wip` when your PR is not ready for review
* **Verification**: The verified label is removed on new commits unless the push is detected as a clean rebase
* **Cherry-picking**: Cherry-pick labels are processed when the PR is merged
* **Container Builds**: Container images are automatically tagged with the PR number
* **Permission Levels**: Some commands require approver permissions
* **Auto-verified Users**: Certain users have automatic verification and merge privileges

### 📌 Additional Information

**Custom Commands:**
- `/test-plan` — Triggers CodeRabbit to analyze the PR's changed files and post a test execution plan
- `/rerun-smoke` — Adds `retest-smoke` label on the PR to trigger smoke tests execution


For more information, please refer to the project documentation or contact the maintainers.

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5759._

<details>
<summary>Overlapping files</summary>

tests/conftest.py

</details>

File: tests/conftest.py [hunk 1]
```diff
@@ -2510,42 +2510,38 @@ def dvs_for_upgrade(
...
 ):
-    golden_images_namespace_name = py_config["golden_images_namespace"]
     dvs_list = []
-    artifactory_secret = utilities.artifactory.get_artifactory_secret(namespace=golden_images_namespace_name)
-    artifactory_config_map = utilities.artifactory.get_artifactory_config_map(namespace=golden_images_namespace_name)
+    with artifactory_credentials(
+        namespace=golden_images_namespace.name, client=golden_images_namespace.client
+    ) as artifactory:
+        for sc in py_config["storage_class_matrix"]:
+            storage_class = [*sc][0]
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/conftest.py:2497]: not the scope of this pr

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5823._

<details>
<summary>Overlapping files</summary>

tests/scale/test_scale_benchmark.py

</details>

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5777._

<details>
<summary>Overlapping files</summary>

tests/storage/conftest.py
tests/utils.py

</details>

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5592._

<details>
<summary>Overlapping files</summary>

tests/infrastructure/golden_images/update_boot_source/test_ssp_data_sources.py

</details>

**Comment by rnetser** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `tests/conftest.py:2508` (outside diff) — MEDIUM/MAJOR: partially-created DataVolumes leak if the setup loop fails mid-way.
> Addressed: Fixed in 2aef6ef3 — wrapped dvs_for_upgrade in ExitStack. Each dv.create() now immediately registers stack.callback(callback=dv.clean_up), so mid-loop failures clean up all previously created DVs. The ExitStack also manages artifactory_credentials, ensuring both DVs and credentials are cleaned up on any failure path.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: utilities/storage.py [hunk 2]
```diff
@@ -244,16 +241,21 @@ def create_dv(
...
+                            client=client,
+                            create_secret=create_secret,
+                            create_config_map=create_config_map,
+                        )
                     )
-                    cert_configmap_name = artifactory_config_map.name
+                    if create_secret:
+                        secret_name = artifactory.secret_name
+                    if create_config_map:
+                        cert_configmap_name = artifactory.cert_configmap_name
```
[the remark this answers is not in the document]
Comment (rnetser) [reply utilities/storage.py:257]: Fixed in 770534aa — updated the create_dv docstring to accurately describe that only caller-unsupplied Secret/ConfigMap resources are created and cleaned up on exit.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by rnetser** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `tests/infrastructure/instance_types/conftest.py:88` (outside diff) — Extract the inline `artifactory_credentials` context into a dedicated fixture for consistency.
> Addressed: Fixed in 770534aa — extracted windows_instance_type_artifactory_credentials fixture in supported_os/conftest.py, consistent with the pattern used across all other conftest files.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

File: tests/conftest.py [hunk 3]
```diff
@@ -14,6 +14,7 @@
 import tempfile
 from bisect import bisect_left
 from collections import defaultdict
+from contextlib import ExitStack
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/conftest.py:17]: Skipped: False positive — all required PR template sections are already present in the PR description with meaningful content: What this PR does, Which issues, Special notes, jira-ticket.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/conftest.py [hunk 4]
```diff
@@ -2498,42 +2499,37 @@ def dvs_for_upgrade(
...
+                    secret_name=artifactory.secret_name,
+                    cert_configmap_name=artifactory.cert_configmap_name,
+                ),
+                storage_class=storage_class,
+                size=rhel_latest_os_params["rhel_dv_size"],
+                annotations=BIND_IMMEDIATE_ANNOTATION,
+                api_name="storage",
+            )
+            dv.create()
+            stack.callback(callback=dv.clean_up)
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/conftest.py]: Fixed in eb69338c — changed to positional form stack.callback(dv.clean_up) and added callback to fcn_exclude_functions in .flake8 to suppress the false-positive FCN001 lint error.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by openshift-virtualization-qe-bot-6** [comment]
**Clean rebase detected** — no code changes compared to previous head (`eb69338`).
The following labels were preserved: `commented-rnetser`, `lgtm-coderabbitai[bot]`.

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5819._

<details>
<summary>Overlapping files</summary>

tests/infrastructure/instance_types/conftest.py

</details>

File: .flake8 [hunk 5]
```diff
@@ -103,6 +103,8 @@ fcn_exclude_functions =
     mkdtemp,
     tempfile,
     urllib3,
+    enter_context,
+    callback,
```
[the remark this answers is not in the document]
Comment (rnetser) [reply .flake8:107]: this is the format that is being used in this repo

File: tests/observability/metrics/utils.py [hunk 6]
```diff
@@ -655,53 +653,47 @@ def create_windows11_wsl2_vm(
...
+                os_flavor=OS_FLAVOR_WINDOWS,
+                name=vm_name,
+                namespace=namespace,
+                client=client,
+                vm_instance_type=VirtualMachineClusterInstancetype(client=client, name="u1.large"),
+                vm_preference=preference,
+                data_volume_template={"metadata": dv.res["metadata"], "spec": dv.res["spec"]},
+            ) as vm:
                 running_vm(vm=vm, dv_wait_timeout=TIMEOUT_40MIN)
                 yield vm
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/observability/metrics/utils.py:698]: not in the scope of this pr

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5834._

<details>
<summary>Overlapping files</summary>

tests/conftest.py

</details>

**Comment by rnetser** [comment]
@EdDev check this out

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5948._

<details>
<summary>Overlapping files</summary>

tests/conftest.py

</details>

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5898._

<details>
<summary>Overlapping files</summary>

tests/conftest.py

</details>

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5813._

<details>
<summary>Overlapping files</summary>

tests/storage/conftest.py
utilities/storage.py
utilities/unittests/test_storage.py

</details>

File: utilities/unittests/test_artifactory.py [hunk 7]
```diff
@@ -332,25 +338,26 @@ def test_get_artifactory_secret_returns_existing_if_exists(self, mock_base64_enc
...
-                get_artifactory_secret(namespace="test-namespace")
+                get_artifactory_secret(namespace="test-namespace", client=MagicMock())
 
     @patch("utilities.artifactory.Secret")
     def test_get_artifactory_secret_raises_key_error_if_token_not_set(self, mock_secret_class):
         """Test raises KeyError if ARTIFACTORY_TOKEN not set"""
         with patch.dict(os.environ, {"ARTIFACTORY_USER": "test-user"}, clear=True):
             with pytest.raises(KeyError):
-                get_artifactory_secret(namespace="test-namespace")
+                get_artifactory_secret(namespace="test-namespace", client=MagicMock())
```
[the remark this answers is not in the document]
Comment (rnetser) [reply utilities/unittests/test_artifactory.py:374]: Removed unused @patch decorators and mock parameters from error-path tests. env var lookup fails before Secret/ConfigMap construction.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: utilities/unittests/test_storage.py [hunk 8]
```diff
@@ -169,3 +170,213 @@ class TestConstructDatavolumeSourceDictUnsupported:
...
+        self,
+        mock_artifactory_credentials,
+        mock_excluded,
+        mock_validate,
+        mock_data_volume_class,
+        mock_wffc,
+    ):
+        credentials = self._credentials()
+        mock_client = MagicMock()
+        mock_artifactory_credentials.side_effect = lambda **kwargs: self._artifactory_cm(credentials)
```
[the remark this answers is not in the document]
Comment (rnetser) [reply utilities/unittests/test_storage.py]: Prefixed unused mock arguments with underscore (_mock_excluded, _mock_validate, _mock_wffc, **_kwargs) across all TestCreateDvArtifactory methods.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/storage/conftest.py [hunk 9]
```diff
@@ -370,19 +370,9 @@ def rhel_vm_name(request):
...
-
-@pytest.fixture(scope="module")
-def artifactory_config_map_scope_module(namespace):
-    artifactory_config_map = get_artifactory_config_map(namespace=namespace.name)
-    yield artifactory_config_map
-    if artifactory_config_map:
-        artifactory_config_map.clean_up()
+def artifactory_credentials_scope_module(namespace):
+    with artifactory_credentials(namespace=namespace.name, client=namespace.client) as credentials:
+        yield credentials
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/storage/conftest.py:375]: False positive — namespace fixture (tests/conftest.py:737) is module-scoped, matching this fixture. No scope mismatch exists.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by rnetser** [comment]
@coderabbitai[bot]

The following review comments were reviewed and a decision was made:

### `utilities/artifactory.py:155` (outside diff) — HIGH: Check resource existence before reading creation-only inputs.
> Addressed: Moved existence check before credential reads. Now creates name-only Secret/ConfigMap for the probe, only reads env vars / SSL cert on the deploy path.
### `tests/scale/test_scale_benchmark.py:253` (outside diff) — HIGH: Credential cleanup can run before DataVolume cleanup — fix the teardown ordering.
> Addressed: Applied ExitStack pattern matching dvs_for_upgrade. Credentials cleanup now runs after DV cleanup (LIFO ordering).
### `utilities/artifactory.py:258` (duplicate) — HIGH: Coordinate overlapping contexts before deleting shared credentials.
> Skipped: Duplicate of prior shared-lifetime discussion. Ownership flags handle the common case — Context B sees pre-existing resource and sets owned=False, so Context A's exit deletes only what it created. The creator-before-borrower edge case is prevented by fixture scope ordering (broader scope outlives narrower).

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/scale/test_scale_benchmark.py [hunk 10]
```diff
@@ -254,44 +251,40 @@ def vms_info(scale_test_param):
...
-
-    cleanup_artifactory_secret_and_config_map(
-        artifactory_secret=artifactory_secret, artifactory_config_map=artifactory_config_map
-    )
+    with ExitStack() as stack:
+        artifactory = stack.enter_context(
+            artifactory_credentials(namespace=golden_images_namespace.name, client=golden_images_namespace.client)
+        )
+        if not keep_resources:
+            stack.callback(delete_resources, resources=dvs_list)
```
[the remark this answers is not in the document]
Comment (rnetser) [reply tests/scale/test_scale_benchmark.py:262]: Added stack.pop_all() before yield when keep_resources=True — detaches all cleanup callbacks so credentials and DVs are retained together.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

[the remark this answers is not in the document]
Comment (rnetser) [reply tests/scale/test_scale_benchmark.py:262]: Moved dvs_list.append() before deploy() so cleanup catches partially-created DVs.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by rnetser** [comment]
@coderabbitai review

**Comment by openshift-virtualization-qe-bot-3** [comment]
/retest all

_Auto-triggered: Files in this PR were modified by merged PR #5672._

<details>
<summary>Overlapping files</summary>

tests/conftest.py

</details>
